src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -249,12 +249,20 @@ fn rustc(cx: &mut Context, unit: &Unit, exec: Arc<Executor>) -> CargoResult<Work
     let mut rustc = prepare_rustc(cx, crate_types, unit)?;
 
     let name = unit.pkg.name().to_string();
+
+    // If this is an upstream dep we don't want warnings from, turn off all
+    // lints.
     if !cx.show_warnings(unit.pkg.package_id()) {
         if cx.config.rustc()?.cap_lints {
             rustc.arg("--cap-lints").arg("allow");
         } else {
             rustc.arg("-Awarnings");
         }
+
+    // If this is an upstream dep but we *do* want warnings, make sure that they
+    // don't fail compilation.
+    } else if !unit.pkg.package_id().source_id().is_path() {
+        rustc.arg("--cap-lints").arg("warn");
```

Comment:
Hm, and do we actually run Cargo's test suite with older compilers? It worries me that this currently passes Travis :)